firmware/system_serial.c

```diff
@@ -21,19 +21,24 @@
 
 #include "system_serial.h"
 
+/*
+ * Read a line from the specified serial connection into the specified
+ * buffer buf with a length of buf_len
+ * This call blocks until a \r or buf_len is reached.
+ */
+size_t serial_getline(SerialDriver *sdp, uint8_t *buf, size_t buf_len)
+{
+    size_t n;
+    uint8_t c;
 
-size_t serial_getline(SerialDriver *sdp, uint8_t *buf, size_t buf_len) {
-  size_t n;
-  uint8_t c;
-
-  n = 0;
-  do {
-    c = sdGet(sdp);
-    *buf++ = c;
-    n++;
-  } while (c != '\r' && n < buf_len - 1);
-  *buf = 0;
-  return n;
+    n = 0;
+    do {
```

Comment:
This loop doesn't handle all conditions.  What if buf was 0 len (Not that it should be, but consider that case)?  This would cause bad behavior (think about the fact that size_t is unsigned here).  I imagine you want this:

```c
if (!buf || !buf_len)
    return 0;

size_t read = 0;
--buf_len; /* account for NULL terminator */
while (read < buf_len) {
    ++read;
    *buf++ = sdGet(sdp);
    if ('\r' == buf[-1])
        break;
}

*buf = '\0';
return read; 
```